mozci/push.py

```diff
@@ -282,8 +282,11 @@ def scheduled_task_labels(self):
         Returns:
             set: A set of task labels (str).
         """
-        tasks = self.decision_task.get_artifact("public/task-graph.json").values()
-        return {t["label"] for t in tasks}
+        try:
+            tasks = self.decision_task.get_artifact("public/task-graph.json").values()
+            return {t["label"] for t in tasks}
+        except Exception as e:
+            return set()
```

Comment:
This seems dangerous. We want this to fail if there is an Exception, otherwise the errors might be going undetected.

Reply:
This is a philosophical difference:  You are assuming you can contain the multitude of problems this code will have in production, and you can handle them all.  I do not care: At this point the artifact does not exist or is corrupted, or the programming is bad; I must proceed with a decision: There are no tasks scheduled for this push.

Reply:
So I was in the process of creating a `MissingArtifactError`, but then realized that the `HTTPError` that requests raises is probably more informative. So why bother catching it and then immediately re-raising? I'll add the fact that `requests.HTTPError` can be raised to the appropriate docstrings and probably call this particular instance a day. I'll see if custom exceptions make more sense in other places.

Reply:
This is not important:

So, `scheduled_task_labels` is an attribute that can throw an `HTTPError`?  This is a low level error given the caller just wants a list of labels.  I do not think the caller should be concerned with **how** this attribute manages to screw up getting a list of labels.  In the same theme, we should not ask the caller to handle http connection setup errors, http download timeouts, or any number of http codes; at some point we bundle many reasons for failure into a single, actionable exception:  `MissingArtifactError` is a good choice.  The caller does not care **why** the artifact can not be had, it only needs to know there is no artifact.

Now, maybe you can set the cause: `MissingArtifactError().__cause__ = HTTPError`; then a smarter caller can still make sophisticated decisions.  Dumb callers can simply mitigate the lack of artifact. 

In general, from a data perspective, there is data or there is not:  In many cases `None` is an appropriate response. In all other cases an exception can be raised; an exception that only describes what the caller was requesting.

```
def get_tasks():
    try:
        # GET TASKS FROM SERVICE, OR CACHE
    except Exception as e:
        raise CanNotGetTasksException(
            "can not get tasks for push {{push}}", 
            push=self.id, 
            cause=e
        )
```

With this pattern, callers can switch behaviour on whether tasks can be had or not (which is most likely), and even switch on causes if they really want (so we never loose the ability to perform fine-grained decision making).

Using this pattern, I have never encountered a method that needs to throw a multiple exception types; multiple reasons are always bundled into one exception indicating failure of the method.

Reply:
Yeah, ok you've convinced me again. That's what I was initially thinking, then started second guessing myself. It's possible to reraise an exception whilst keeping the original stacktrace which seems like what we'd want here.

Reply:
I believe python3 has better causal support.  Long ago I built libraries to keep stacktraces and causes for python2, and I have never looked into what Python3 is capable of now.  Sorry I am not much actual help.